waiter/src/waiter/scheduler.clj

```diff
@@ -638,3 +642,145 @@
           "WAITER_PASSWORD" (service-id->password-fn service-id)
           "WAITER_SERVICE_ID" service-id
           "WAITER_USERNAME" "waiter"}))
+
+;; Support for tracking service instance launching time stats
+
+(def instance-counts-zero {:requested 0 :started 0})
+
+(defn- make-launch-tracker-entry
+  [service-id]
+  [service-id
+   {:instance-counts instance-counts-zero
+    :known-instance-ids #{}
+    :instance-scheduling-start-times []
+    :service-schedule-timer (metrics/service-timer service-id "launch-overhead" "schedule-time")
+    :service-startup-timer (metrics/service-timer service-id "launch-overhead" "startup-time")
+    :starting-instance-ids []}])
+
+(defn update-launch-trackers
+  "Track metrics on app instances, specifically schedule and startup times."
+  [service-id->launch-tracker new-service-ids removed-service-ids service-id->healthy-instances
+   service-id->unhealthy-instances service-id->instance-counts waiter-schedule-timer]
+  (let [now (t/now)
+        new-tracker-entries (map make-launch-tracker-entry new-service-ids)
+        tracker-entries' (->> service-id->launch-tracker
+                              ;; Remove trackers from deleted services
+                              (remove (comp removed-service-ids key))
+                              ;; Add trackers for newly-discovered services
+                              (concat new-tracker-entries))]
+    ;; Update all trackers...
+    (pc/for-map [[service-id
+                  {:keys [known-instance-ids instance-counts instance-scheduling-start-times
+                          service-schedule-timer service-startup-timer starting-instance-ids]
+                   :as tracker-state}]
+                 tracker-entries']
+      service-id
+      (let [healthy-instances (get service-id->healthy-instances service-id)
+            known-instances' (->> service-id
+                                  (get service-id->unhealthy-instances)
+                                  (into healthy-instances))
```

Comment:
Should we worry about the growth of this vector of `known-instances`, e.g. if a service scales up and down over its lifetime, we will end up storing all its instances ever created? We could choose to say not remember instances that were started over an hour ago (based on our max supported grace period).

Reply:
My understanding is that making this set the union of the current *healthy* and *unhealthy* instance sets means that this only includes currently-live instances. If we don't have a set of currently-live instances, then we don't have a good way to determine when new instances appear.

Reply:
For some reason I believed `known-instances'` was aggregating over the previous `known-instances`. Since we are not doing that, we will not be storing all instances ever created and the concern is no longer valid. Also, there is no `known-instances` in this context, should we rename `known-instances'` to `known-instances`?

Reply:
Done.